walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,124 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.PriorityQueue;
+import java.util.Set;
 
 public final class FindMeetingQuery {
+  private static final Event ENTIRE_DAY_EVENT = new Event("Entire Day Event", TimeRange.WHOLE_DAY, Arrays.asList());
+
+  /** Returns a Collection of TimeRanges that satisfy the details of the MeetingRequest based on the events */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<String> requiredAttendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+
+    // Initialize PriorityQueue with start and end 'EventTime' nodes
+    PriorityQueue<EventTime> eventTimes = new PriorityQueue<>();
+    for (Event event : events) {
+      TimeRange eventTimeRange = event.getWhen();
+      eventTimes.add(new EventTime(eventTimeRange.start(), event, true));
+      eventTimes.add(new EventTime(eventTimeRange.end(), event, false));
+    }
+    // Add EventTime nodes to the beginning and end of the day, to handle edge cases
+    eventTimes.add(new EventTime(ENTIRE_DAY_EVENT.getWhen().start(), ENTIRE_DAY_EVENT, true));
+    eventTimes.add(new EventTime(ENTIRE_DAY_EVENT.getWhen().end(), ENTIRE_DAY_EVENT, false));
+
+    // Keep track of potential meetings and the Attendees that are 'busy' while going through the PriorityQueue
+    Set<Event> ongoingEvents = new HashSet<Event>();
+    List<TimeRange> possibleMeetings = new ArrayList<>();
+    List<TimeRange> possibleMeetingsWithOptAttendees = new ArrayList<>();
+
+    // 'Traverse' throughout the whole day, one EventTime at a time,
+    // and find potential blocks of time for Meetings
+    EventTime previousEventTime = eventTimes.poll();
+    ongoingEvents.add(previousEventTime.getEvent());
+    while (!eventTimes.isEmpty()) {
+      EventTime currentEventTime = eventTimes.poll();
+
+      // Check if any ongoing event conflicts with the optional or required attendees
+      boolean meetingConflictsReqAttendees = false;
+      boolean meetingConflictsOptAttendees = false;
+      for (Event event : ongoingEvents) {
+        for (String attendee : event.getAttendees()) {
+          if (requiredAttendees.contains(attendee)) {
+            meetingConflictsReqAttendees = true;
+            meetingConflictsOptAttendees = true;
```

Comment:
Conceptually, I don't like this line. `requiredAttendees` are not optional, so why are we marking a conflict for optional attendees?